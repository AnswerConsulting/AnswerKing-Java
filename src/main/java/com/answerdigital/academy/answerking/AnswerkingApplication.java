package com.answerdigital.academy.answerking;

import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;

@SpringBootApplication
public class AnswerkingApplication {
    public static void main(final String[] args) {
	SpringApplication.run(AnswerkingApplication.class, args);
    }
}
