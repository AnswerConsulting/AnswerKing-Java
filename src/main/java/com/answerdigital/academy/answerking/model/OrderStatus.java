package com.answerdigital.academy.answerking.model;

public enum OrderStatus {
    SUCCESSFUL,
    CANCELLED,
    IN_PROGRESS
}
