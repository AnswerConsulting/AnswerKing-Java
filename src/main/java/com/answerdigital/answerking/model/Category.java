package com.answerdigital.answerking.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.Column;
import javax.persistence.FetchType;
import javax.persistence.OneToMany;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.Pattern;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

import static com.answerdigital.answerking.util.DateTimeUtility.getDateTimeAsString;

@Entity
@Getter
@Setter
@Builder
@AllArgsConstructor
@NoArgsConstructor(access = AccessLevel.PROTECTED, force = true)
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Pattern(regexp = "^[a-zA-Z\s-]*", message = "Category name must only contain letters, spaces and dashes")
    private String name;

    @NotBlank
    @Pattern(regexp = "^[a-zA-Z\s.,!?0-9-']*",
            message = "Category description can only contain letters, numbers, spaces and !?-.,' punctuation")
    private String description;

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @Column(name = "created_on")
    private String createdOn;

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    @Column(name = "last_updated")
    private String lastUpdated;

    private boolean retired;

    @OneToMany(mappedBy="category", fetch = FetchType.EAGER)
    @JsonIgnore
    private Set<Product> products = new HashSet<>();

    public Category(final String name, final String description) {
        this.name = name;
        this.description = description;
        this.retired = false;
        this.createdOn = getDateTimeAsString();
        this.lastUpdated = this.createdOn;
    }

    public void addProduct(final Product product) {
        products.add(product);
    }

    public void removeProduct(final Product product) {
        products.remove(product);
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final Category category = (Category) o;
        return name.equals(category.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return "Category{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                '}';
    }
}
