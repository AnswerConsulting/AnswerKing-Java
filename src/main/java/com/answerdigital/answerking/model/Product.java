package com.answerdigital.answerking.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToMany;
import javax.persistence.OneToMany;
import javax.persistence.PreRemove;
import javax.persistence.Table;
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Table(name = "product")
@JsonIgnoreProperties({"hibernateLazyInitializer"})
@Getter
@Setter
@Builder
@NoArgsConstructor(access = AccessLevel.PROTECTED, force = true)
@AllArgsConstructor
public class Product {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private Long id;

    private String name;

    private String description;

    @Column(precision = 12, scale = 2)
    private BigDecimal price;

    @JsonProperty(access = JsonProperty.Access.READ_ONLY)
    private boolean retired;

    @JsonIgnore
    @ManyToMany(fetch = FetchType.LAZY, mappedBy = "products")
    private Set<Category> categories = new HashSet<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL)
    @JsonIgnore
    private Set<LineItem> lineItems = new HashSet<>();

    public Product(final String name, final String description, final BigDecimal price, final boolean isRetired) {
        this.name = name;
        this.description = description;
        this.price = price;
        this.categories = new HashSet<>();
        this.retired = isRetired;
    }

    @PreRemove
    private void removeProductsFromCategories() {
        categories.forEach(category -> category.removeProduct(this));
    }

    public BigDecimal getPrice() {
        return price.setScale(2, RoundingMode.DOWN);
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final Product product = (Product) o;
        return Objects.equals(id, product.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return "Product{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", price=" + price +
                ", retired=" + retired +
                '}';
    }
}
