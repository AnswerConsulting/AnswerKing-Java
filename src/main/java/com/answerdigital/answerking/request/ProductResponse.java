package com.answerdigital.answerking.request;

public class ProductResponse {



}
