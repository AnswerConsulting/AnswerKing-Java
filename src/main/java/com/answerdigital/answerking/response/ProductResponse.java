package com.answerdigital.answerking.response;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.util.List;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class ProductResponse {

    private Long id;

    private String name;

    private String description;

    private BigDecimal price;

    private List<Long> categories;

    private boolean retired;
}
