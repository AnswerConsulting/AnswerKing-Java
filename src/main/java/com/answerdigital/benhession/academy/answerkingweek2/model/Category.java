package com.answerdigital.benhession.academy.answerkingweek2.model;

import com.answerdigital.benhession.academy.answerkingweek2.dto.AddCategoryDTO;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.OneToMany;
import javax.validation.constraints.NotNull;
import java.util.HashSet;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

@Entity
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private  int id;
    @NotNull
    private String name;
    @NotNull
    private String description;
    @OneToMany(mappedBy = "category")
    private final Set<ItemCategory> itemCategories = new HashSet<>();

    public Category(String name, String description) {
        this.name = name;
        this.description = description;
    }

    public Category(AddCategoryDTO addCategoryDTO) {
        this.name = addCategoryDTO.getName();
        this.description = addCategoryDTO.getDescription();
    }

    public Category() {

    }

    void addItemCategory(ItemCategory itemCategory) {
         itemCategories.add(itemCategory);
    }

    void removeItem(Item item) {

        Optional<ItemCategory> itemCategory = itemCategories.stream()
                .filter(ic -> ic.getItem().equals(item))
                .findFirst();

        itemCategory.map(itemCategories::remove);

    }

    public boolean containsItem(Item item) {
        return getItems().contains(item);
    }

    public int getId() {
        return id;
    }

    public Set<Item> getItems() {
        return itemCategories.stream().map(ItemCategory::getItem).collect(Collectors.toSet());
    }

    public Set<ItemCategory> getItemCategories() {
        return new HashSet<>(itemCategories);
    }

    public void clearItemCategories() {
        itemCategories.clear();
    }

    public String getName() {
        return name;
    }

    public String getDescription() {
        return description;
    }

    public void setName(String name) {
        this.name = name;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public void setId(int id) {
        this.id = id;
    }


    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Category category = (Category) o;
        return name.equals(category.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    @Override
    public String toString() {
        return "Category{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", itemCategories=" + itemCategories +
                '}';
    }
}
