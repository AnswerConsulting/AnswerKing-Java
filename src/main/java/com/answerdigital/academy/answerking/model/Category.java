package com.answerdigital.academy.answerking.model;

import com.answerdigital.academy.answerking.request.AddCategoryRequest;
import com.fasterxml.jackson.annotation.JsonIgnore;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import lombok.AccessLevel;

import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.Pattern;
import java.util.HashSet;
import java.util.Objects;
import java.util.Set;

@Entity
@Getter
@Setter
@Builder
@AllArgsConstructor
@NoArgsConstructor(access = AccessLevel.PROTECTED, force = true)
public class Category {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotBlank
    @Pattern(regexp = "^[a-zA-Z\s-]*", message = "Category name must only contain letters, spaces and dashes")
    private String name;

    @NotBlank
    @Pattern(regexp = "^[a-zA-Z\s.,!?0-9-']*",
            message = "Category description can only contain letters, numbers, spaces and !?-.,' punctuation")
    private String description;

    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "item_category",
            joinColumns = {@JoinColumn(name = "category_id")},
            inverseJoinColumns = {@JoinColumn(name = "item_id")})
    @JsonIgnore
    private Set<Item> items = new HashSet<>();

    public Category(final AddCategoryRequest categoryRequest) {
        this.name = categoryRequest.name();
        this.description = categoryRequest.description();
    }

    public void addItem(final Item item) {
        items.add(item);
    }

    public void removeItem(final Item item) {
        items.remove(item);
    }

    public Category(final String name, final String description) {
        this.name = name;
        this.description = description;
    }

    @Override
    public boolean equals(final Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        final Category category = (Category) o;
        return name.equals(category.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return "Category{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                '}';
    }
}
