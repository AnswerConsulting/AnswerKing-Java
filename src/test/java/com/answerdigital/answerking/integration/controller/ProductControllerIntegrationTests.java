package com.answerdigital.answerking.integration.controller;

import com.answerdigital.answerking.AnswerKingApplication;
import com.answerdigital.answerking.model.Category;
import com.answerdigital.answerking.model.Product;
import com.answerdigital.answerking.response.ProductResponse;
import com.answerdigital.answerking.utility.AbstractContainerBaseTest;
import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.api.Tag;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.context.jdbc.Sql;
import org.springframework.test.web.servlet.MockMvc;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
import static org.assertj.core.api.Assertions.assertThat;

@SpringBootTest(classes = {AnswerKingApplication.class, Category.class, Product.class, ProductResponse.class},
                webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
@AutoConfigureMockMvc(addFilters = false)
@ActiveProfiles("integration-test")
@ExtendWith({SnapshotExtension.class})
class ProductControllerIntegrationTests extends AbstractContainerBaseTest {

    private Expect expect;
    public String uuidRegex = "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}";

    @Autowired
    private MockMvc mvc;

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getAllProductsTest() throws Exception {

        final var response = mvc.perform(get("/products"))
            .andExpect(status().isOk())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response);
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql"})
    void getAllProductsEmptyTest() throws Exception {

        final var response = mvc.perform(get("/products"))
            .andExpect(status().isNoContent())
            .andReturn()
            .getResponse()
            .getContentAsString();

        assertThat(response).isEqualTo("[]");
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getProductByIDTest() throws Exception {

        final var response = mvc.perform(get("/products/1")) // Need to get ID from seed really
            .andExpect(status().isOk())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response);
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getProductByIDNotFoundTest() throws Exception {

        final var response = mvc.perform(get("/products/10000")) 
            .andExpect(status().isNotFound())
            .andReturn()
            .getResponse()
            .getContentAsString();

        String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        expect.serializer("json").toMatchSnapshot(sanitisedResponse); // This could be handled much better if the json was serialised. Then we could point to the json field and sanatise using that.
    }

    @Test
    @Tag("GET")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void getProductByIDInvalidTest() throws Exception {

        final var response = mvc.perform(get("/products/invalid")) 
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        expect.serializer("json").toMatchSnapshot(sanitisedResponse); // This could be handled much better if the json was serialised. Then we could point to the json field and sanatise using that.
    }

    @Test
    @Tag("POST")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void postProductTest() throws Exception {

        final var response = mvc.perform(post("/products")
            .contentType("application/json")
            .content("{\"name\": \"Kabab\", \"description\": \"Product 1 desc\", \"price\": 10.34, \"categoryId\": 1}")) 
            // Need to add static fixtures instead of writing this in tests
            .andExpect(status().isCreated())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response); 
    }

    @Test
    @Tag("POST")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void postProductTestInvalidBody() throws Exception {

        final var response = mvc.perform(post("/products")
            .contentType("application/json")
            .content("This is not valid JSON")) 
            // Need to add static fixtures and be data driven to check all different ways body can be incorrect
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        expect.serializer("json").toMatchSnapshot(sanitisedResponse); // This could be handled much better if the json was serialised. Then we could point to the json field and sanatise using that.
    }

    @Test
    @Tag("PUT")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void putProductTest() throws Exception {

        final var response = mvc.perform(put("/products/2")
            .contentType("application/json")
            .content("{\"name\": \"Kabab\", \"description\": \"Product 1 desc\", \"price\": 10.34, \"categoryId\": 1}")) 
            // Need to add static fixtures instead of writing this in tests
            .andExpect(status().isOk())
            .andReturn()
            .getResponse()
            .getContentAsString();

        expect.serializer("json").toMatchSnapshot(response); 
    }

    @Test
    @Tag("PUT")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void putProductTestInvalidBody() throws Exception {

        final var response = mvc.perform(put("/products/2")
            .contentType("application/json")
            .content("This is not valid JSON")) 
            // Need to add static fixtures and be data driven to check all different ways body can be incorrect
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

            String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
            expect.serializer("json").toMatchSnapshot(sanitisedResponse); // This could be handled much better if the json was serialised. Then we could point to the json field and sanatise using that.
    }

    @Test
    @Tag("PUT")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void putProductByIDInvalidTest() throws Exception {

        final var response = mvc.perform(put("/products/invalid")) 
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        expect.serializer("json").toMatchSnapshot(sanitisedResponse); // This could be handled much better if the json was serialised. Then we could point to the json field and sanatise using that.
    }

    @Test
    @Tag("DELETE")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void deleteProductByIDTest() throws Exception {

        mvc.perform(delete("/products/2")) // Need to get ID from seed really
            .andExpect(status().isNoContent())
            .andReturn()
            .getResponse()
            .getContentAsString();
        // Add another assertion that checks the db and makes sure retired is set to true
    }

    @Test
    @Tag("DELETE")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void deleteProductByIDNotFoundTest() throws Exception {

        final var response = mvc.perform(delete("/products/10000")) 
            .andExpect(status().isNotFound())
            .andReturn()
            .getResponse()
            .getContentAsString();

        String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        expect.serializer("json").toMatchSnapshot(sanitisedResponse); // This could be handled much better if the json was serialised. Then we could point to the json field and sanatise using that.
    }

    @Test
    @Tag("DELETE")
    @Sql({"/test_sql_scripts/clearAll4Test.sql", "/test_sql_scripts/product_controller_test_input.sql"})
    void deleteProductByIDInvalidTest() throws Exception {

        final var response = mvc.perform(delete("/products/invalid")) 
            .andExpect(status().isBadRequest())
            .andReturn()
            .getResponse()
            .getContentAsString();

        String sanitisedResponse = response.replaceAll(uuidRegex, "{ Sanitised }");
        expect.serializer("json").toMatchSnapshot(sanitisedResponse); // This could be handled much better if the json was serialised. Then we could point to the json field and sanatise using that.
    }
}
