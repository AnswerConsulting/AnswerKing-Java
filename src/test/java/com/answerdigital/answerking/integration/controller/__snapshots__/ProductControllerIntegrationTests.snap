com.answerdigital.answerking.integration.controller.ProductControllerIntegrationTests.getAllProductsTest=[
  "[{\"id\":1,\"name\":\"Burger\",\"description\":\"300g Beef\",\"price\":6.69,\"category\":{\"id\":1,\"name\":\"CatOne\",\"description\":\"CatOne description\"},\"retired\":false},{\"id\":2,\"name\":\"Cola\",\"description\":\"500ml\",\"price\":2.99,\"category\":{\"id\":2,\"name\":\"CatTwo\",\"description\":\"CatTwo description\"},\"retired\":false},{\"id\":3,\"name\":\"Fries\",\"description\":\"Large Fries\",\"price\":2.99,\"category\":{\"id\":1,\"name\":\"CatOne\",\"description\":\"CatOne description\"},\"retired\":false}]"
]